File: problems/1657/Lc1657.ts
function char2num(ch: string) {
  return ch.charCodeAt(0) - 97;
};

function closeStrings(word1: string, word2: string): boolean {
  if (word1.length !== word2.length)
    return false;

  const freq1 = new Array<number>(26).fill(0),
    freq2 = new Array<number>(26).fill(0);
  const present1 = new Set<string>(),
    present2 = new Set<string>();
  for (const w of word1) {
    freq1[char2num(w)]++;
    present1.add(w);
  }
  for (const w of word2) {
    freq2[char2num(w)]++;
    present2.add(w);
  }

  if (present1.size !== present2.size || ![...present1].every((w) => (present2.has(w))))
    return false;

  freq1.sort((l, r) => (l - r));
  freq2.sort((l, r) => (l - r));
  return freq1.every((val, pos) => (val === freq2[pos]));
};

function main() {
  let word1: string, word2: string, result: boolean, expected: boolean;

  word1 = 'abc';
  word2 = 'bca';
  result = closeStrings(word1, word2);
  expected = true;
  expected === result || console.trace(`expected ${expected}, but ${result}`);

  word1 = 'a';
  word2 = 'aa';
  result = closeStrings(word1, word2);
  expected = false;
  expected === result || console.trace(`expected ${expected}, but ${result}`);

  word1 = 'cabbba';
  word2 = 'abbccc';
  result = closeStrings(word1, word2);
  expected = true;
  expected === result || console.trace(`expected ${expected}, but ${result}`);

  word1 = 'abbzccca';
  word2 = 'babzzczc';
  result = closeStrings(word1, word2);
  expected = true;
  expected === result || console.trace(`expected ${expected}, but ${result}`);

  word1 = 'aaabbbbccddeeeeefffff';
  word2 = 'aaaaabbcccdddeeeeffff';
  result = closeStrings(word1, word2);
  expected = false;
  expected === result || console.trace(`expected ${expected}, but ${result}`);

  word1 = 'uau';
  word2 = 'ssx';
  result = closeStrings(word1, word2);
  expected = false;
  expected === result || console.trace(`expected ${expected}, but ${result}`);
};
main();

File: problems/1657/Lc1657.c
#include <assert.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

int desc(const void* larg, const void* rarg) {
  int lval = *(const int*)larg, rval = *(const int*)rarg;
  return (lval < rval) - (lval > rval);
}

bool closeStrings(char* word1, char* word2) {
  size_t word1size = strlen(word1), word2size = strlen(word2);
  if (word1size != word2size)
    return false;

  int freq1[26] = {0}, freq2[26] = {0};
  for (size_t pos = 0; pos < word1size; ++pos)
    ++freq1[word1[pos] - 'a'];
  for (size_t pos = 0; pos < word2size; ++pos)
    ++freq2[word2[pos] - 'a'];
  for (size_t pos = 0; pos < 26; ++pos) {
    if ((freq1[pos] == 0) != (freq2[pos] == 0))
      return false;
  }
  qsort(freq1, 26, sizeof(int), desc);
  qsort(freq2, 26, sizeof(int), desc);
  for (size_t pos = 0; pos < 26; ++pos) {
    if (freq1[pos] != freq2[pos])
      return false;
    if (freq1[pos] == 0 && freq2[pos] == 0)
      break;
  }
  return true;
}

void testcase01() {
  char* word1 = "abc";
  char* word2 = "bca";
  bool result = closeStrings(word1, word2);
  bool expected = true;
  assert(expected == result);
}

void testcase02() {
  char* word1 = "a";
  char* word2 = "aa";
  bool result = closeStrings(word1, word2);
  bool expected = false;
  assert(expected == result);
}

void testcase03() {
  char* word1 = "cabbba";
  char* word2 = "abbccc";
  bool result = closeStrings(word1, word2);
  bool expected = true;
  assert(expected == result);
}

void testcase04() {
  char* word1 = "abbzccca";
  char* word2 = "babzzczc";
  bool result = closeStrings(word1, word2);
  bool expected = false;
  assert(expected == result);
}

void testcase05() {
  char* word1 = "aaabbbbccddeeeeefffff";
  char* word2 = "aaaaabbcccdddeeeeffff";
  bool result = closeStrings(word1, word2);
  bool expected = false;
  assert(expected == result);
}

void testcase06() {
  char* word1 = "uau";
  char* word2 = "ssx";
  bool result = closeStrings(word1, word2);
  bool expected = false;
  assert(expected == result);
}

int main() {
  testcase01();
  testcase02();
  testcase03();
  testcase04();
  testcase05();
  testcase06();
  return 0;
}
